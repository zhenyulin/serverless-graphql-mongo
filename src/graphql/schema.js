import schemaComposer from './resolver';

export default schemaComposer.buildSchema();
