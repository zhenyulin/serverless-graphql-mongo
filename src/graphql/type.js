export default `
	type User {
		_id: ID!
		name: String
		email: String!
		followees: [User]
		followers: [User]
		likeItems: [Item]
	}

	input UserInput {
		name: String
		email: String
	}

	type Item {
		_id: ID!
		name: String!
		likedByUsers: [User]
	}

	input ItemInput {
		name: String
	}

	type Query {
		User(_id: ID!): User
		Users: [User]
		Item(_id: ID!): Item
		Items: [Item]
	}

	type Mutation {
		CreateUser(user: UserInput): User
		UpdateUser(_id: ID!, user: UserInput): User
		DeleteUser(_id: ID!): User
		CreateItem(item: ItemInput): Item
		UpdateItem(_id: ID!, item: ItemInput): Item
		DeleteItem(_id: ID!): Item
		UserFollowUser(followerId: ID!, followeeId: ID!): User
		UserUnfollowUser(followerId: ID!, followeeId: ID!): User
		UserLikeItem(userId: ID!, itemId: ID!): User
		UserDisplikeItem(userId: ID!, itemId: ID!): User
	}
`;
