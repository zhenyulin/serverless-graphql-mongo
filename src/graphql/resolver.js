import Mongo from '../models';

export default {
	Query: {
		User: (_, { _id }) => Mongo.User.findById(_id),
		Users: () => Mongo.User.find(),
		Item: (_, { _id }) => Mongo.Item.findById(_id),
		Items: () => Mongo.Item.find(),
	},
	User: {
		followees: async ({ _id }) => {
			const userFollows = await Mongo.UserFollowUser.find({
				followerId: _id,
			});
			const followeeIds = userFollows.map(follow => follow.followeeId);
			const followees = await Mongo.User.find({ _id: { $in: followeeIds } });
			return followees;
		},
		followers: async ({ _id }) => {
			const userFollowed = await Mongo.UserFollowUser.find({
				followeeId: _id,
			});
			const followerIds = userFollowed.map(follow => follow.followerId);
			const followers = await Mongo.User.find({ _id: { $in: followerIds } });
			return followers;
		},
		likeItems: async ({ _id }) => {
			const likes = await Mongo.UserLikeItem.find({
				userId: _id,
			});
			const itemIds = likes.map(like => like.itemId);
			const items = await Mongo.Item.find({ _id: { $in: itemIds } });
			return items;
		},
	},
	Item: {
		likedByUsers: async ({ _id }) => {
			const likes = await Mongo.UserLikeItem.find({
				itemId: _id,
			});
			const userIds = likes.map(like => like.userId);
			const users = await Mongo.User.find({ _id: { $in: userIds } });
			return users;
		},
	},
	Mutation: {
		CreateUser: (_, { user }) => Mongo.User.create(user),
		UpdateUser: (_, { _id, user }) => Mongo.User.findByIdAndUpdate(_id, user),
		DeleteUser: (_, { _id }) => Mongo.User.findOneAndDelete(_id),
		CreateItem: (_, { item }) => Mongo.Item.create(item),
		UpdateItem: (_, { _id, item }) => Mongo.Item.findByIdAndUpdate(_id, item),
		DeleteItem: (_, { _id }) => Mongo.Item.findOneAndDelete(_id),
		UserFollowUser: async (_, { followerId, followeeId }) => {
			await Mongo.UserFollowUser.create({
				followerId,
				followeeId,
			});
			const user = await Mongo.User.findById(followerId);
			return user;
		},
		UserUnfollowUser: async (_, { followerId, followeeId }) => {
			await Mongo.UserFollowUser.findOneAndDelete({ followerId, followeeId });
			const user = await Mongo.User.findById(followerId);
			return user;
		},
		UserLikeItem: async (_, { userId, itemId }) => {
			await Mongo.UserLikeItem.create({
				userId,
				itemId,
			});
			const user = await Mongo.User.findById(userId);
			return user;
		},
		UserDisplikeItem: async (_, { userId, itemId }) => {
			await Mongo.UserFollowUser.findOneAndDelete({ userId, itemId });
			const user = await Mongo.User.findById(userId);
			return user;
		},
	},
};
