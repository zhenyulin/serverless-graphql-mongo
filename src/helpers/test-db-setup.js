import mongoose from 'mongoose';

import DB from 'models';

import * as mock from '__fixtures__';

export const connect = () =>
	mongoose.connect(
		process.env.MONGODB_URI_TEST,
		{
			useNewUrlParser: true,
		},
	);

export const disconnect = () => mongoose.disconnect();

export const loadUsers = () => DB.User.create(mock.USERS);
export const loadItems = () => DB.Item.create(mock.ITEMS);
export const loadUserFollowUsers = () =>
	DB.UserFollowUser.create(mock.USER_FOLLOW_USERS);
export const loadUserLikeItems = () =>
	DB.UserLikeItem.create(mock.USER_LIKE_ITEMS);

export const unloadUsers = () => DB.User.remove();
export const unloadItems = () => DB.Item.remove();
export const unloadUserFollowUsers = () => DB.UserFollowUser.remove();
export const unloadUserLikeItems = () => DB.UserFollowUser.remove();

export const loadAll = () =>
	Promise.all([
		loadUsers(),
		loadItems(),
		loadUserFollowUsers(),
		loadUserLikeItems(),
	]);

export const unloadAll = () =>
	Promise.all([
		unloadUsers(),
		unloadItems(),
		unloadUserFollowUsers(),
		unloadUserLikeItems(),
	]);
