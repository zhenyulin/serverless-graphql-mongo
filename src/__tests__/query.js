import { graphql } from 'graphql';

import schema from 'graphql/schema';

import * as testDB from './helpers/test-db-setup';

const testCases = cases => {
	Object.keys(cases).forEach(caseName => {
		it(caseName, async () => {
			const query = cases[caseName];
			const result = await graphql(schema, query);
			expect(result).toMatchSnapshot();
		});
	});
};

describe('Query', () => {
	beforeAll(async () => {
		await testDB.connect();
		await testDB.unloadAll();
		await testDB.loadAll();
	});

	afterAll(async () => {
		await testDB.disconnect();
	});

	describe('User', () => {
		testCases({
			'get user by _id': `
				query {
					User(_id: "4edd40c86762e0fb12000003") {
						_id
						name
						followees {
							name
						}
						followers {
							name
						}
						likeItems {
							name
						}
					}
				}
			`,
			'return null if user not found': `
				query {
					User(_id: "4edd40c86762e0fb12000099") {
						_id
						name
					}
				}
			`,
			'throw error if _id is not provided': `
				query {
					User() {
						_id
						name
					}
				}
			`,
			'throw error if _id is not valid mongodb id': `
				query {
					User(_id: "random") {
						_id
						name
					}
				}
			`,
		});
	});

	describe('Users', () => {
		testCases({
			'get all users': `
				query {
					Users {
						_id
						name
						followees {
							name
						}
						followers {
							name
						}
						likeItems {
							name
						}
					}
				}
			`,
		});
	});

	describe('Item', () => {
		testCases({
			'get item by _id': `
				query {
					Item(_id: "4edd40c86762e0fb12000015") {
						_id
						name
						likedByUsers {
							name
						}
					}
				}
			`,
			'return null if item not found': `
				query {
					Item(_id: "4edd40c86762e0fb12000099") {
						_id
						name
					}
				}
			`,
			'throw error if _id is not provided': `
				query {
					Item() {
						_id
						name
					}
				}
			`,
			'throw error if _id is not valid mongodb id': `
				query {
					Item(_id: "random") {
						_id
						name
					}
				}
			`,
		});
	});

	describe('Items', () => {
		testCases({
			'get all items': `
				query {
					Items {
						_id
						name
						likedByUsers {
							name
						}
					}
				}
			`,
		});
	});
});
