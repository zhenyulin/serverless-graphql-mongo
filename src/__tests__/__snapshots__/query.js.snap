// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query Item get item by _id 1`] = `
Object {
  "data": Object {
    "Item": Object {
      "_id": "4edd40c86762e0fb12000015",
      "likedByUsers": Array [],
      "name": "Birth of Cool",
    },
  },
}
`;

exports[`Query Item return null if item not found 1`] = `
Object {
  "data": Object {
    "Item": null,
  },
}
`;

exports[`Query Item throw error if _id is not provided 1`] = `
Object {
  "errors": Array [
    [GraphQLError: Syntax Error: Expected Name, found )],
  ],
}
`;

exports[`Query Item throw error if _id is not valid mongodb id 1`] = `
Object {
  "data": Object {
    "Item": null,
  },
  "errors": Array [
    [GraphQLError: Cast to ObjectId failed for value "random" at path "_id" for model "Item"],
  ],
}
`;

exports[`Query Items get all items 1`] = `
Object {
  "data": Object {
    "Items": Array [
      Object {
        "_id": "4edd40c86762e0fb12000013",
        "likedByUsers": Array [],
        "name": "Imagine",
      },
      Object {
        "_id": "4edd40c86762e0fb12000014",
        "likedByUsers": Array [
          Object {
            "name": "John Lennon",
          },
        ],
        "name": "4:00",
      },
      Object {
        "_id": "4edd40c86762e0fb12000015",
        "likedByUsers": Array [],
        "name": "Birth of Cool",
      },
    ],
  },
}
`;

exports[`Query User get user by _id 1`] = `
Object {
  "data": Object {
    "User": Object {
      "_id": "4edd40c86762e0fb12000003",
      "followees": Array [],
      "followers": Array [
        Object {
          "name": "John Lennon",
        },
      ],
      "likeItems": Array [],
      "name": "John Cage",
    },
  },
}
`;

exports[`Query User return null if user not found 1`] = `
Object {
  "data": Object {
    "User": null,
  },
}
`;

exports[`Query User throw error if _id is not provided 1`] = `
Object {
  "errors": Array [
    [GraphQLError: Syntax Error: Expected Name, found )],
  ],
}
`;

exports[`Query User throw error if _id is not valid mongodb id 1`] = `
Object {
  "data": Object {
    "User": null,
  },
  "errors": Array [
    [GraphQLError: Cast to ObjectId failed for value "random" at path "_id" for model "User"],
  ],
}
`;

exports[`Query Users get all users 1`] = `
Object {
  "data": Object {
    "Users": Array [
      Object {
        "_id": "4edd40c86762e0fb12000003",
        "followees": Array [],
        "followers": Array [
          Object {
            "name": "John Lennon",
          },
        ],
        "likeItems": Array [],
        "name": "John Cage",
      },
      Object {
        "_id": "4edd40c86762e0fb12000004",
        "followees": Array [
          Object {
            "name": "John Cage",
          },
        ],
        "followers": Array [],
        "likeItems": Array [
          Object {
            "name": "4:00",
          },
        ],
        "name": "John Lennon",
      },
      Object {
        "_id": "4edd40c86762e0fb12000005",
        "followees": Array [],
        "followers": Array [],
        "likeItems": Array [],
        "name": "Miles Davis",
      },
    ],
  },
}
`;
